### DISCUSSION TITLE
Extending auth providers

### DISCUSSION THREAD

**Discussion by DraftProducts** [adonisjs/core#1765] [General]
Hi,
I looked to the documentation to find how to extend **auth providers** but i can't find it anywere.
I know that the documentation is in progress so if someone can give me some indications or help to build and register a custom provider **I could with pleasure offer a pull request to explain this concept in the documentation**.

For the moment i juste saw the extend method in the [AuthManager class](https://github.com/adonisjs/auth/blob/develop/src/AuthManager/index.ts#L252-L267).

Firstly i thought i just had to create a Provider in the [providers folder](https://preview.adonisjs.com/guides/directory-structure#the-providers-directory) but the schema requested in not the same as the Auth provider.

If someone have allready created a custom guard, i will happy to see it as example.

Thank you for the help you can provide to this discussion !

**Comment by thetutlage** [comment]
If I get it right you are trying to create a custom guard right?

**Reply by DraftProducts** [reply]
You're right !
I try to understand how i can extend the Guard to override the methods and change how the auth is checked and work.

**Reply by DraftProducts** [reply]
In my case, i would like to have real users, and temporary users
A real user have a registration and a informations, (name, username, email, localisation) and a temporary user is just a username linked to a token (auth token). So if the user want to save his progression (at the end of the work, lesson, questions) he can register and this user will be linked. The goal is that in a group i have temporary users and reals users.
To get this, i have to change how the auth system know if the user is auth but also what information is return for the user, because if there is no user linked to the token i have to return a user object with only a username.

**Reply by DraftProducts** [reply]
Or more simply add a boolean to the token table to find out if the user id belongs to the users table or to the temporary users table.

**Reply by thetutlage** [reply]
I don't think you need to implement separate driver all together for this. Simply use 2 different guards with 2 different user providers. The config may look something like this

```ts
users: {
    driver: 'session',
    provider: {
      driver: 'lucid',
      identifierKey: 'id',
      uids: ['email'],
      model: User,
    },
},

guests: {
    driver: 'session',
    provider: {
      driver: 'lucid',
      identifierKey: 'id',
      uids: ['token'],
      model: Guest,
    },
},
```

And then at runtime you can switch between the guards `auth.use('users')` or `auth.use('guests')`

As a side note:

- You create guard drivers when you are implementing a different authentication mode in your app. For example: Authenticate using sessions, or tokens or JWT or some 3rd party service
- You create providers when you want to fetch users from a different source, like some 3rd party API

**Reply by DraftProducts** [reply]
Okay, thanks a lot @thetutlage !
I think I'm going in that direction.
So i don't need to extend auth providers anymore, but can you please explain how for the topic ?
For futur users witch have this question "how extend auth providers ?", so il will mark this answer as reply.

**Reply by thetutlage** [reply]
This will need proper documentation and explaining it in a post is kind of hard. I have plans to proper docs around extending AdonisJS, but time is limited, so will get to them sooner.

**Reply by DraftProducts** [reply]
I know that's a lot of work, can i help with something ?

**Reply by thetutlage** [reply]
Depends upon your interests. We need a lot of help with documentation, articles and small recipes. In terms of code, I would love if someone can pick one of the following tasks

- Wrapping socket.io to work seamlessly with AdonisJS. This includes making the AdonisJS middleware work.
- Generate models from an existing database. This needs working with the TypeScript AST using packages like [recast](https://www.npmjs.com/package/recast)

Feel free to send me DM  on Discord and I can help further in getting started

**Reply by DraftProducts** [reply]
Oh nice, I'm currently working on a project with Socket.io so that would be a pleasure :)
https://github.com/Bio2Game/Bio2Game-API
For the moment i just used the AppProvider

**Comment by fairhat** [comment]
@DraftProducts 

Let's say you want to implement a custom TeamGuard for your Sports related website and on some routes people can only view them if they belong to the team they want to view:

```ts
import { HttpContextContract } from '@ioc:Adonis/Core/HttpContext'
import NotYourTeamException from 'App/Exceptions/NotYourTeamException'

export default class TeamGuard {
  public async handle (ctx: HttpContextContract, next: () => Promise<void>) {
    const { auth, params } = ctx
    await auth.user!.preload('teams')
    const teamId = params.teamId
    const team = auth.user?.teams.find(team => team.id === teamId)
    const doesntBelongToTeam = !team

    if (doesntBelongToTeam) {
      throw new NotYourTeamException ()
    }

    ctx.team = team
    await next()
  }
}
```
This guard assumes that you will only use it together with the default AuthGuard AND only after the authguard since it relies on "auth.user" to exist.
This would check which teams this user belongs to and if this user belongs to the requested team (inside the route params (':teamId')).
As long as users aren't assigned to hundreds of teams this should be efficient enough.

Now we register this guard as a middleware inside the kernel.ts:

```ts
Server.middleware.registerNamed({
  auth: 'App/Middleware/Auth',
  team: 'App/Middleware/TeamGuard',
})
```

Now we can use them for the routes:
```ts
Route.get('/team/:teamId', 'TeamController.getTeam').middleware([ 'auth', 'team' ])
```

**Reply by DraftProducts** [reply]
Good evening, thank you for your answer @fairhat, unfortunately I was not talking about middleweare but about auth providers, witch is not the same thing. I would like to change how the auth/tokens are checked and what informations are send.
